### PR TITLE
improve target coordinate specification in jplhorizons

### LINKED ISSUE #2599: Topographic position disallowed in vectors query but would be useful for tracking
With these example definitions
```
DSN_goldstone = '257@399'
SSN_millstone = {'lon':-71.490967, 'lat':42.617442, 'elevation':0.127}
epochs = {'start': '1998-01-23 06:00:00', 'stop': '1998-01-23 12:00:00', 'step': '1m'}
spacecraft='NEAR'
```
the call `Horizons(id=spacecraft, location=DSN_goldstone, epochs=epochs).vectors()` returns topocentric range and range rate representative of Doppler tracking from DSN Goldstone in this case. The equivalent call `Horizons(id=spacecraft, location=SSN_millstone, epochs=epochs).vectors()`, to compare radar range from MIT's Haystack is refused with `ValueError: cannot use topographic position in statevectors query`, solely because of its location argument format.

This seems inconsistent, and perhaps the first query should be also blocked for violating the spirit of the vectors query. However, I feel the Horizons topocentric range and range rate would be a reliable reference for cross-checking radar or telemetry tracking of satellites.

### PR DESCRIPTION
**Pull request by m-stclair**
closes https://github.com/astropy/astroquery/issues/1555
Fixes #2599.

### CODE REVIEW

**Comment by m-stclair** [comment]
hi @bsipocz -- I've reimplemented these features, added basic test coverage for them, updated other tests as necessary, and updated docstrings. I'm assuming the fast CI failures where the pipeline can't find `tox` are because this is a draft; if not, I have no idea how to interrogate those. What else do you think needs to be done before I mark this as ready for review?

**Comment by bsipocz** [comment]
It's a weird and new failure about tox, the draft status of the PR shouldn't affect the CI this way. I'll look into it and fix it separately.

If the content is ready, then do mark it ready for review, and we can ask @mkelley to do a first round of review while I'm dealing with CI.
(Tests can be checked locally, including the remote tests, etc.)

**Comment by bsipocz** [comment]
@m-stclair - could you rebase? CI has been fixed, but I cannot step in as a maintainer to do the rebase as the PR was opened from an org fork rather than from yours.

**Comment by m-stclair** [comment]
Ty for fixing the tox issue. I have rebased and am now marking this ready for review.

File: astroquery/jplhorizons/tests/test_jplhorizons.py [hunk 1]
```diff
@@ -256,22 +256,23 @@ def test_vectors_query_payload():
     res = jplhorizons.Horizons(id='Ceres', location='500@10',
                                epochs=2451544.5).vectors(
                                    get_query_payload=True)
-
     assert res == OrderedDict([
-        ('format', 'text'),
```
Comment (bsipocz) [inline astroquery/jplhorizons/tests/test_jplhorizons.py:261]: Please refrain from reformatting like this. Add the new dict element, but leave the rest alone.

File: astroquery/jplhorizons/core.py [hunk 2]
```diff
@@ -1181,14 +1224,23 @@ def _parse_result(self, response, verbose=None):
...
                 headerline[2] = 'solar_presence'
-                headerline[3] = 'flags'
+                if 'Earth' in centername:
+                    headerline[3] = 'lunar_presence'
+                else:
+                    headerline[3] = 'interfering_body'
                 headerline[-1] = '_dump'
+                if (
+                    isinstance(self.id, dict) or str(self.id).startswith('g:')
+                ):
```
Comment (bsipocz) [inline astroquery/jplhorizons/core.py:1241]: We have longer than 80 char lines, do keep this in one line instead of 1 with information and 2 of fillers. This is for here and a few other places, too.

File: astroquery/jplhorizons/core.py [hunk 3]
```diff
@@ -1060,18 +1075,17 @@ def vectors_async(self, *, get_query_payload=False,
...
             if no_fragments:
                 commandline += ' NOFRAG;'
-
-        if isinstance(self.location, dict):
-            raise ValueError(('cannot use topographic position in state'
-                              'vectors query'))
+        #
+        # if isinstance(self.location, dict):
+        #     raise ValueError(('cannot use topographic position in state'
+        #                       'vectors query'))
```
Comment (bsipocz) [inline astroquery/jplhorizons/core.py:1081]: Do remove unused code rather than comment out

File: astroquery/jplhorizons/core.py [hunk 4]
```diff
@@ -580,19 +601,9 @@ def ephemerides_async(self, *, airmass_lessthan=99,
...
-                self.location['body'] = '399'
-            request_payload['CENTER'] = 'coord@{:s}'.format(
-                str(self.location['body']))
-            request_payload['COORD_TYPE'] = 'GEODETIC'
-            request_payload['SITE_COORD'] = "'{:f},{:f},{:f}'".format(
-                self.location['lon'], self.location['lat'],
-                self.location['elevation'])
+            request_payload = dict(
+                **request_payload, **self._location_to_params(self.location)
+            )
```
Comment (bsipocz) [inline astroquery/jplhorizons/core.py:606]: keep it in one line

File: astroquery/jplhorizons/core.py [hunk 5]
```diff
@@ -15,6 +12,9 @@
 from astropy.time import Time
 from astropy.utils.exceptions import AstropyDeprecationWarning
 from astropy.utils.decorators import deprecated_renamed_argument, deprecated_attribute
+from numpy import nan
```
Comment (bsipocz) [inline astroquery/jplhorizons/core.py:15]: if you need to reorder imports, keep the numpy ones before astropy.

File: astroquery/jplhorizons/core.py [hunk 6]
```diff
@@ -1032,17 +1043,21 @@ def vectors_async(self, *, get_query_payload=False,
...
 
-        # check for required information
+        # check for required information and assemble commandline stub
         if self.id is None:
             raise ValueError("'id' parameter not set. Query aborted.")
+        elif isinstance(self.id, dict):
+            commandline = (
+                f"g:{self.id['lon']},{self.id['lat']},"
+                f"{self.id['elevation']}@{self.id['body']}"
+            )
```
Comment (bsipocz) [inline astroquery/jplhorizons/core.py:1053]: this also probably fit the line lenght limit

**Review by bsipocz — CHANGES_REQUESTED** [review]
Some minor nitpicks, and please add a changelog entry.  

Also, please update the narrative docs, changes here have effect on some of the doctest expectations there.

File: astroquery/jplhorizons/tests/test_jplhorizons.py (hunk 1 above)
Comment (m-stclair) [reply astroquery/jplhorizons/tests/test_jplhorizons.py:261]: Switching the order back will cause the test to fail, because it is an equality test against an `OrderedDict`, and changes I need to make in order to construct the payload to include target coordinates alter the order in which keys are inserted in the `OrderedDict`.

File: astroquery/jplhorizons/core.py (hunk 2 above)
Comment (m-stclair) [reply astroquery/jplhorizons/core.py:1241]: copy that.

File: astroquery/jplhorizons/core.py (hunk 3 above)
Comment (m-stclair) [reply astroquery/jplhorizons/core.py:1081]: oops. missed this.

File: astroquery/jplhorizons/core.py (hunk 4 above)
Comment (m-stclair) [reply astroquery/jplhorizons/core.py:606]: copy that.

File: astroquery/jplhorizons/core.py (hunk 5 above)
Comment (m-stclair) [reply astroquery/jplhorizons/core.py:15]: copy that.

File: astroquery/jplhorizons/core.py (hunk 6 above)
Comment (m-stclair) [reply astroquery/jplhorizons/core.py:1053]: condensed to one line.

File: astroquery/jplhorizons/tests/test_jplhorizons.py (hunk 1 above)
Comment (m-stclair) [reply astroquery/jplhorizons/tests/test_jplhorizons.py:261]: will add a changelog entry and update narrative docs tomorrow or tuesday.

Comment (bsipocz) [reply astroquery/jplhorizons/tests/test_jplhorizons.py:261]: I don't mean the order, but don't do reformatting, no need to change 15 lines when only one or two need to be shuffled around

**Comment by m-stclair** [comment]
will add a changelog entry and update narrative docs tomorrow or tuesday.

**Comment by bsipocz** [comment]
> The codecov misses are pretty small things, but let me know if it's important to cover them and I'll add a couple of little unit tests!

codecov is red herring here, it doesn't look for the changes with the remote tests, so it misses the actual ~2% test increase

**Comment by m-stclair** [comment]
on codecov: got it, makes sense

File: astroquery/jplhorizons/tests/test_jplhorizons.py (hunk 1 above)
Comment (m-stclair) [reply astroquery/jplhorizons/tests/test_jplhorizons.py:261]: oh you mean the indentation. sorry, didn't even notice.

**Comment by m-stclair** [comment]
@bsipocz , I believe I've addressed all your requests.

File: astroquery/jplhorizons/core.py [hunk 7]
```diff
@@ -1032,17 +1041,18 @@ def vectors_async(self, *, get_query_payload=False,
 
         URL = conf.horizons_server
 
-        # check for required information
+        # check for required information and assemble commandline stub
         if self.id is None:
             raise ValueError("'id' parameter not set. Query aborted.")
+        elif isinstance(self.id, dict):
+            commandline = f"g:{self.id['lon']},{self.id['lat']},{self.id['elevation']}@{self.id['body']}"
```
Comment (eerovaher) [inline astroquery/jplhorizons/core.py:1048]: Usually I prefer f-strings over `str.format()`, but in this case the following can be considered:
```suggestion
            commandline = "g:{lon},{lat},{elevation}@{body}".format(**self.id)
```

File: astroquery/jplhorizons/core.py [hunk 8]
```diff
@@ -1132,6 +1141,30 @@ def vectors_async(self, *, get_query_payload=False,
         return response
 
     # ---------------------------------- parser functions
+    @staticmethod
+    def _prep_loc_dict(loc_dict, attr_name):
+        """prepare coord specification dict for 'location' or 'id'"""
+        if {'lat', 'lon', 'elevation'} - set(loc_dict.keys()) != set():
```
Comment (eerovaher) [inline astroquery/jplhorizons/core.py:1147]: An empty set is falsy and a non-empty set is truthy. And the explicit conversion of dict keys to `set` is not needed (at least in Python 3.10):
```suggestion
        if {'lat', 'lon', 'elevation'} - loc_dict.keys():
```

File: astroquery/jplhorizons/core.py [hunk 9]
```diff
@@ -1181,14 +1214,21 @@ def _parse_result(self, response, verbose=None):
...
             # read in ephemerides header line; replace some field names
             if (self.query_type == 'ephemerides' and
                     "Date__(UT)__HR:MN" in line):
                 headerline = str(line).split(',')
                 headerline[2] = 'solar_presence'
-                headerline[3] = 'flags'
+                if 'Earth' in centername:
+                    headerline[3] = 'lunar_presence'
+                else:
+                    headerline[3] = 'interfering_body'
```
Comment (eerovaher) [inline astroquery/jplhorizons/core.py:1227]: Is the one-liner too long?
```suggestion
                    headerline[3] = "lunar_presence" if "Earth" in centername else "interfering_body"
```

File: astroquery/jplhorizons/tests/test_jplhorizons_remote.py [hunk 10]
```diff
@@ -405,3 +409,51 @@ def test_ephemerides_extraprecision(self):
...
+
+    def test_geodetic_queries(self):
+        """
+        black-box test for observer and vectors queries with geodetic
+        coordinates. checks spatial sensibility.
+        """
+        phobos = {
+            'body': 401,
+            'lon': random.randint(-150, -1),
+            'lat': random.randint(-80, 80),
```
Comment (eerovaher) [inline astroquery/jplhorizons/tests/test_jplhorizons_remote.py:421]: I'm not sure running the tests on random inputs is worth the effort. Even if the tests stumble upon a weird edge case that causes the tests to fail, it can be very difficult to figure out what the problematic values were, which prevents them from being addressed.

File: docs/jplhorizons/jplhorizons.rst [hunk 11]
```diff
@@ -72,6 +105,24 @@ as the observer's location:
...
+at a particular point in time to the center of the crater Double:
+
+.. doctest-remote-data::
+
+    >>> ce_2 = {'lon': 23.522, 'lat': 0.637, 'elevation': 181.2, 'body': 301}
+    >>> double = {'lon': 23.47, 'lat': 0.67, 'elevation': 0, 'body': 301}
+    >>> obj = Horizons(id=double, location=ce_2, epochs=2454483.84247)
+    >>> vecs = obj.vectors()
+    >>> distance_km = (vecs['x'] ** 2 + vecs['y'] ** 2 + vecs['z'] ** 2) ** 0.5 * 1.496e8
+    >>> print(round(distance_km.value.data[0], 3))
```
Comment (eerovaher) [inline docs/jplhorizons/jplhorizons.rst:124]:
```suggestion
    >>> print(f"{distance_km.value.data[0]:.3f}")
```

File: astroquery/jplhorizons/core.py (hunk 8 above)
Comment (m-stclair) [reply astroquery/jplhorizons/core.py:1147]: Huh, I didn't know that about `dict_keys`, neat -- set operators in general fail with non-set iterables and I'd never tried. 

I have an allergy to truthiness tests but I understand that people often prefer them. Is your preference here strong?

File: astroquery/jplhorizons/core.py (hunk 9 above)
Comment (m-stclair) [reply astroquery/jplhorizons/core.py:1227]: it was too long for what I thought the line length limit was when I was writing it -- most of the module had been wrapped at 79 -- but I have since learned that this is not a real `astroquery` restriction and that longer lines are in fact preferred. So this would work.

File: astroquery/jplhorizons/core.py (hunk 7 above)
Comment (m-stclair) [reply astroquery/jplhorizons/core.py:1048]: this is prettier, yes.

File: docs/jplhorizons/jplhorizons.rst (hunk 11 above)
Comment (m-stclair) [reply docs/jplhorizons/jplhorizons.rst:124]: better, yes

File: astroquery/jplhorizons/tests/test_jplhorizons_remote.py (hunk 10 above)
Comment (m-stclair) [reply astroquery/jplhorizons/tests/test_jplhorizons_remote.py:421]: It makes me nervous to use only white-box tests, but I understand if you prefer it for stability reasons. So I will turn this into another explicit equality test.

File: astroquery/jplhorizons/core.py (hunk 8 above)
Comment (m-stclair) [reply astroquery/jplhorizons/core.py:1147]: Actually, my preference isn't _that_ strong. I'll just change this.

File: astroquery/jplhorizons/tests/test_jplhorizons_remote.py (hunk 10 above)
Comment (eerovaher) [reply astroquery/jplhorizons/tests/test_jplhorizons_remote.py]: The best way of doing randomized tests would be to use [`hypothesis`](https://hypothesis.readthedocs.io/en/latest/) because it is capable of storing inputs that caused a failure, but `astroquery` isn't using it at the moment.

**Comment by m-stclair** [comment]
@eerovaher , I believe I've addressed all your suggestions. I do not understand why the readthedocs build is now failing, but it does not appear to be related to any actual code changes.

File: astroquery/jplhorizons/tests/test_jplhorizons_remote.py (hunk 10 above)
Comment (m-stclair) [reply astroquery/jplhorizons/tests/test_jplhorizons_remote.py]: I do like `hypothesis`. But this is totally fine with me. Weird edge cases are not actually that likely here.

**Comment by eerovaher** [comment]
I don't really understand why the documentation build failed. Perhaps if you squash some of the commits (not necessarily down to one) and then force-push to GitHub the next documentation build might succeed.

**Comment by bsipocz** [comment]
> I do not understand why the readthedocs build is now failing, but it does not appear to be related to any actual code changes.

This is a weird problem, I don't think I saw it before (the commit hashes match, yet RTD complains that it cannot check it out). A restart didn't resolve it, so I would say wait it out while I investigate whether other repos are affected, too, or try with the squashing idea.

**Comment by bsipocz** [comment]
OK, leave the RTD for now, this is a more widespread problem and not much we can do about it in this PR.

**Comment by earthshrink** [comment]
hi @m-stclair, in first comment above, did you mean closes #2599?

**Comment by m-stclair** [comment]
No, it's just the confusing GitHub convention where PRs and issues are notated the same way, and this is PR 2625. When I opened this PR I hadn't noticed you'd opened that issue. However, it's definitely true that this closes #2599.

**Comment by bsipocz** [comment]
Sorry, that was my bad, I indeed meant to add https://github.com/astropy/astroquery/issues/2599 there and not the PR's number. Thanks for spotting it. 😄

**Comment by m-stclair** [comment]
Nevermind then! I thought it had been automatically added.

**Comment by bsipocz** [comment]
Yes, your original PR predates all the new issues. And based on that, this would also address https://github.com/astropy/astroquery/issues/1555

**Review by bsipocz — APPROVED** [review]

**Comment by bsipocz** [comment]
Thanks @m-stclair!
